### DISCUSSION TITLE
How can I build a Rest API to return all class names?

### DISCUSSION THREAD

**Discussion by lvsouza** [tailwindlabs/tailwindcss#5086] [Help] [answered]
Hi, I have a web application that needs to list to the end user the list of all css classes available in tailwindcss. However, I know very well that the classes are generated according to a configuration file.

The idea is to pass the configuration in the body of a post.

**My problem**: How can I generate the classes in the backend based on the configuration passed?

If anyone can give me any tips on how to do this, I would be very grateful. Thank you!

**Comment by dance2die** [comment]
> **My problem**: How can I generate the classes in the backend based on the configuration passed?

I don't know how to generate classes at least I can point to where to get started. 

You can resolve the configuration and access the configuration: https://tailwindcss.com/docs/configuration#referencing-in-java-script

**Comment by simonswiss** [comment]
Hey!

Unfortunately there is no public APIs to recreate what Tailwind does to generate the classes based on the config file, but... if you want to learn what happens behind the scenes, you can take a look at the "chain of transforms" applied here...

https://github.com/tailwindlabs/tailwindcss/blob/master/src/processTailwindFeatures.js#L67-L77

```js
    postcss([
      substituteTailwindAtRules(...),
      evaluateTailwindFunctions(...),
      substituteVariantsAtRules(...),
      substituteResponsiveAtRules(...),
      convertLayerAtRulesToControlComments(...),
      substituteScreenAtRules(...),
      substituteClassApplyAtRules(...),
      applyImportantConfiguration(...),
      purgeUnusedStyles(...),
    ])
```

... and reverse engineer your way from there 😅

All these "steps" are located in the [`src/lib` folder](https://github.com/tailwindlabs/tailwindcss/tree/master/src/lib), and reading through that code is a big eye-opener!

I bet you wish there was a `generateTailwindOutput(config)` function you could import from `tailwindcss/generate` and run it... but there is not.

To be honest you're probably better off building simple functions that take your theme, and for each property (fontSize, spacing, ...), iterate over the values and output the correct class for each.

That in itself is a fun JavaScript exercise if you want to practice iterating over `Object.entries`, using `Array.map()`, filter etc 🎉

And let me tell you, it's really rewarding when you get it right and you see all the values show up on your screen!

I bet this would be a fun little project using NextJS, which supports [API routes](https://nextjs.org/docs/api-routes/introduction), which you could use to create your API.

Hope it helps you getting going!

**Comment by lvsouza** [comment, marked as answer]
@simonswiss @dance2die thank you very much for the replies!

I did a little research, and I developed a way to do this.

I don't think this is the best and most performative way, but it will suit me!

I put the code in the repository below, in case anyone is interested.
https://github.com/lvsouza/tailwind-generator

**Comment by lesaff** [comment]
I am also interested in generating a json list of all available class names. I found this JS package that generates PHP array from tailwindjs config file. https://github.com/vendeka-nl/tailwind-config-php

I am planning to write a function to generate class names from the output per @simonswiss' suggestion above.
